### DISCUSSION TITLE
Motion planning with obstacle avoidance

### DISCUSSION THREAD

**Discussion by janneyow** [bulletphysics/bullet3#4027] [Q&A] [answered]
Hi, I'm using calculateInverseKinematics to plan the trajectory of a 6-dof robot arm but the arm collides with the other objects in the environment. Is there a built-in function to perform motion planning with obstacle avoidance?

Thanks!

**Comment by rohit-kumar-j** [comment, marked as answer]
@janneyow, no there isn't, you'll have to design/generate the path with some logic yourself.

**Comment by BluePrintRandom** [comment]
check this out
here is A* and flood fill search and broadphase search

https://www.redblobgames.com/pathfinding/a-star/implementation.py
https://www.redblobgames.com/pathfinding/a-star/implementation.html

when we evaluate paths, we can calculate the closest point using a KDTree or bvhtree of the obstacles.

you can run a simulation of a robots 'plan' and calculate the closest point per joint to obstacles and adjust the plan or abandon it and try and generate a new one (gan?)

**Reply by rohit-kumar-j** [reply]
@BluePrintRandom this is a good explanation of the implementation of search algorithms!

**Comment by BluePrintRandom** [comment]
with a IK creature - I actually do a planning phase per 'step' - maybe during this phase you can raytest / adjust the foot arc
**
https://www.youtube.com/watch?v=fMwe9hGZb9w
